tests/integration/nm/dns_test.py

```diff
@@ -304,3 +304,45 @@ def test_global_dns_do_not_touch_iface_dns(static_iface_dns):
         "nmcli -g ipv4.dns c show eth1".split(), check=True
     )[1]
     assert output.strip() == ",".join(original_dns_servers)
+
+
+# Adapted reproducer for https://issues.redhat.com/browse/RHEL-31095
+def test_keep_interface_dns_when_not_changing_global_dns():
+    try:
+        # Add and activate a connection that has a nameserver set
+        cmdlib.exec_cmd(
+            f"nmcli connection delete {DUMMY0}".split(), check=False
+        )
+        cmdlib.exec_cmd(
+            f"nmcli connection add type dummy con-name {DUMMY0} ifname {DUMMY0} "
+            "autoconnect no ipv4.method manual ipv4.address 172.31.43.1/30 "
+            "ipv4.dns 127.0.53.53".split(),
+            check=True,
+        )
+        cmdlib.exec_cmd(f"nmcli connection up {DUMMY0}".split(), check=True)
+
+        # Apply changes to connection that are not related to DNS
+        libnmstate.apply(
+            {
+                Interface.KEY: [
+                    {
+                        Interface.NAME: DUMMY0,
+                        Interface.TYPE: InterfaceType.DUMMY,
+                        Interface.STATE: InterfaceState.UP,
+                        Interface.IPV6: {
+                            InterfaceIPv6.ENABLED: False,
+                        },
+                    },
+                ],
+            }
+        )
+
+        # Check the nameserver is still there
+        output = cmdlib.exec_cmd(
+            f"nmcli -t -f ipv4.dns c show {DUMMY0}".split(), check=True
+        )[1]
+        assert "ipv4.dns:127.0.53.53" in output
+    finally:
+        cmdlib.exec_cmd(
+            f"nmcli connection delete {DUMMY0}".split(), check=False
+        )
```

Comment:
We may move the clean code to fixture `static_iface_dns` like:

```python
    yield
    libnmstate.apply(
        {
            DNS.KEY: {DNS.CONFIG: {DNS.SERVER: [], DNS.SEARCH: []}},
        }
    )
```

Reply:
Done